policies/codingstyle.txt

```diff
@@ -289,12 +289,78 @@ options.
 
 In source files, separate functions with one blank line.
 
+                Chapter 6.1: Function Arguments
+
+Most API functions in OpenSSL revolve around a specific type (e.g. SSL_CTX,
+EVP_PKEY, X509 etc). When adding new API functions relevant to a given type, the
+following argument ordering should be used:
+
+    rettype TYPE_get_something(TYPE *,[args])
+    rettype TYPE_set_something(TYPE *,[args])
+    rettype TYPE_do_something(TYPE *,[args])
+    TYPE *TYPE_new([args])
+
+Some more complex complex APIs have a TYPE and TYPE_CTX pair, where TYPE usually
+represents the set of function pointers to the implementation, while TYPE_CTX is
+the collection of attributes to perform the intended operation.
+For that case, the pattern should also include:
+
+    TYPE_CTX *TYPE_CTX_new([args])
+    rettype TYPE_CTX_get_something(TYPE_CTX *,[args])
+    rettype TYPE_CTX_set_something(TYPE_CTX *,[args])
+    rettype TYPE_do_something(TYPE_CTX *,[args])
+
+Where present the TYPE (or TYPE_CTX) argument should always be first as shown
+in the examples above. In constructors or object factory type functions (such as
+a "fetch" function), the key argument that identifies the thing be constructed
+should be placed first (if such an argument exists). For example in a "fetch"
+function the name of the algorithm being fetched would be first.
```

Comment:
I had glossed over this section...  I disagree in so far that if there is another context that acts as a factory pool for implicit fetches and similar, it actually makes sense to have it first, regardless of if those implicit fetches happen directly or later.  So in a constructor that takes a `OPENSSL_CTX *`, I would place that first.

The idea is also that we *don't* view libctx + propq as a pair, but rather that propq is associated with the algorithm to be fetched, which means that the fetching functions as they currently are fine, as well as, for example, this:

``` C
EVP_PKEY_CTX *EVP_PKEY_CTX_new_from_name(OPENSSL_CTX *libctx,
                                         const char *name,
                                         const char *propquery);
EVP_PKEY_CTX *EVP_PKEY_CTX_new_from_pkey(OPENSSL_CTX *libctx,
                                         EVP_PKEY *pkey, const char *propquery);
```

Ref: https://github.com/openssl/openssl/blob/4f14a378f807e989aa0b328267732409c8d6ac68/include/openssl/evp.h#L1594-L1598

I did a write-up on this last week, see https://mta.openssl.org/pipermail/openssl-project/2020-September/002160.html
And yeah, I did change my mind compared to what I've said before that.  Decoupling libctx from propq made me rethink.

Reply:
> And yeah, I did change my mind compared to what I've said before that. Decoupling libctx from propq made me rethink.

FFS. This clause was there to keep you happy!! :-) Ok - I'll remove it.

It's like herding cats around here.

Reply:
I'm sorry. We were obviously not in sync